### DISCUSSION TITLE
What do these errors really mean?

### DISCUSSION THREAD

**Discussion by marcelstoer** [khoih-prog/EthernetWebServer_SSL#19] [Q&A]
I have difficulties assessing the severity of errors reported by the library. The HTTPS exchange does succeed, no problem, but...

What is the system trying to tell me? Should I be worried? What should be done differently?

```
23:18:11.190 > Received 255847 bytes
23:18:12.157 > (EthernetSSLClient)(SSL_ERROR)(m_update_engine): Error writing to m_client
23:18:12.163 > (EthernetSSLClient)(SSL_ERROR)(m_update_engine): 0
23:18:12.168 > (EthernetSSLClient)(SSL_WARN)(connected): Socket was dropped unexpectedly (this can be an alternative to closing the connection)
23:18:12.180 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
23:18:12.186 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
23:18:12.190 > (EthernetSSLClient)(SSL_ERROR)(available): SSL engine failed to update.
23:18:12.196 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
23:18:12.201 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
23:18:12.212 > Total byte count: 268316
23:18:12.212 > 
23:18:12.212 > Disconnecting. Status code: 4
23:18:12.218 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
23:18:12.223 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
23:18:12.229 > Received 268316 bytes in 21.8135 s, rate = 12.30 kbytes/second
```

**Comment by khoih-prog** [comment]
1.

> ```
> 23:18:12.157 > (EthernetSSLClient)(SSL_ERROR)(m_update_engine): Error writing to m_client
> 23:18:12.163 > (EthernetSSLClient)(SSL_ERROR)(m_update_engine): 0
> ```

From

https://github.com/khoih-prog/EthernetWebServer_SSL/blob/15a1d4c925a82cfff324179b1ef94ff2002073f1/src/SSLClient/SSLClient_Impl.h#L660-L667

Nothing to worry because `!m_client.connected()` as `m_client.getWriteError() == 0`

**We'd tried to write while client is not connected**

2. The followings are the results of 1., 

```
23:18:12.168 > (EthernetSSLClient)(SSL_WARN)(connected): Socket was dropped unexpectedly (this can be an alternative to closing the connection)
23:18:12.180 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
23:18:12.186 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
23:18:12.190 > (EthernetSSLClient)(SSL_ERROR)(available): SSL engine failed to update.
23:18:12.196 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
23:18:12.201 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
23:18:12.212 > Total byte count: 268316
23:18:12.212 > 
23:18:12.212 > Disconnecting. Status code: 4
23:18:12.218 > (EthernetSSLClient)(SSL_ERROR)(connected): Not connected because write error is set
23:18:12.223 > (EthernetSSLClient)(SSL_ERROR)(m_print_ssl_error): SSL_CLIENT_WRITE_FAIL
```
 **We'd tried to write while client is not connected**

So check the code to see where, then fix if any issue. Also check **to be sure client is connected before writing something.**

Possibly a bug somewhere in the library ??? or your code ??? I don't know

I don't think this can create any big issue because we'll rewrite if some data / packets are lost. But it's still better to find out the root cause.

As you're familiar with [**SSLClient library**](https://github.com/OPEnSLab-OSU/SSLClient), you can also try to duplicate and make a [**Minimal, Reproducible Example**](https://stackoverflow.com/help/minimal-reproducible-example) and post an issue/ discussion there.

**Reply by marcelstoer** [reply]
After some more digging it appears to be this bug here: https://github.com/OPEnSLab-OSU/SSLClient/issues/43
